zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/SendAckRequestProcessor.java

```diff
@@ -64,7 +66,8 @@ public void flush() throws IOException {
         } catch (IOException e) {
             LOG.warn("Closing connection to leader, exception during packet send", e);
             try {
-                if (!learner.sock.isClosed()) {
+                Socket socket = learner.sock;
+                if ( socket != null && ! learner.sock.isClosed()) {
```

Comment:
Should probably use `socket` in the second condition too? In case it changes after the first check.

Reply:
Thanks, that was of course the intention :) Fixed!

Reply:
hi jonmv, i have read the jira-4541. I am confusing. ZK1 does not send ack to  leader , ZK1 recieves commit from leader. It seems not conform ZAB protocol. Please help me figure out, thanks